File: src/main/java/no/nav/veilarboppfolging/domain/OppfolgingStatusData.java
package no.nav.veilarboppfolging.domain;

import lombok.Data;
import lombok.experimental.Accessors;
import no.nav.veilarboppfolging.repository.entity.EskaleringsvarselEntity;
import no.nav.veilarboppfolging.repository.entity.KvpEntity;
import no.nav.veilarboppfolging.repository.entity.OppfolgingsperiodeEntity;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static java.util.Comparator.naturalOrder;

@Data
@Accessors(chain = true)
public class OppfolgingStatusData {
    public String fnr;
    public String aktorId;
    public String veilederId;
    public boolean reservasjonKRR;
    public boolean manuell;
    public boolean underOppfolging;
    public boolean underKvp;
    public boolean kanStarteOppfolging;
    public boolean kanVarsles;
    private EskaleringsvarselEntity gjeldendeEskaleringsvarsel;
    public List<OppfolgingsperiodeEntity> oppfolgingsperioder = Collections.emptyList();
    public List<KvpEntity> kvpPerioder;
    public boolean harSkriveTilgang;
    public Boolean inaktivIArena;
    public Boolean kanReaktiveres;
    public LocalDate inaktiveringsdato;
    public Boolean erSykmeldtMedArbeidsgiver;
    public String servicegruppe;
    public String formidlingsgruppe;
    public String rettighetsgruppe;

    @Deprecated
    public Boolean erIkkeArbeidssokerUtenOppfolging;

    public ZonedDateTime getOppfolgingUtgang() {
        return oppfolgingsperioder.stream().map(OppfolgingsperiodeEntity::getSluttDato).filter(Objects::nonNull).max(naturalOrder()).orElse(null);
    }

}

File: src/main/kotlin/no/nav/veilarboppfolging/domain/OppfolgingStatusData.kt
package no.nav.veilarboppfolging.domain

import no.nav.veilarboppfolging.repository.entity.KvpPeriodeEntity
import no.nav.veilarboppfolging.repository.entity.OppfolgingsperiodeEntity
import java.time.LocalDate
import java.time.ZonedDateTime
import java.util.*

data class OppfolgingStatusData(
    val fnr: String,
    val aktorId: String,
    val veilederId: String? = null,
    val reservasjonKRR: Boolean,
    val registrertKRR: Boolean,
    val manuell: Boolean,
    val underOppfolging: Boolean,
    val underKvp: Boolean,
    val kanStarteOppfolging: Boolean?,
    val kanVarsles: Boolean,
    val oppfolgingsperioder: List<OppfolgingsperiodeEntity>,
    val kvpPerioder: List<KvpPeriodeEntity>,
    val harSkriveTilgang: Boolean,
    val inaktivIArena: Boolean?,
    val kanReaktiveres: Boolean?,
    val inaktiveringsdato: LocalDate?,
    val erSykmeldtMedArbeidsgiver: Boolean?,
    val servicegruppe: String?,
    val formidlingsgruppe: String?,
    val rettighetsgruppe: String?,
    @Deprecated("")
    val erIkkeArbeidssokerUtenOppfolging: Boolean? = null
) {
    val oppfolgingUtgang: ZonedDateTime?
        get() = oppfolgingsperioder.stream().map<ZonedDateTime?>(OppfolgingsperiodeEntity::sluttDato)
            .filter { obj: ZonedDateTime? -> Objects.nonNull(obj) }
            .max(Comparator.naturalOrder()).orElse(null)
}